### PR TITLE
Add type safety for spool state and desired spool state

### PR DESCRIPTION
**Pull request by peterbarker**
- canonicalises on "spool state" rather than "spool mode"
- state "spool up" becomes state "spooling up"
- fixed mixed use of enumerations:
```
../../ArduCopter/mode.cpp:638:14: fatal error: comparison of two values with different enumeration types in switch statement ('enum spool_up_down_mode' and 'AP_Motors::spool_up_down_desired') [-Wenum-compare-switch]
        case AP_Motors::DESIRED_GROUND_IDLE:
```
- used switch statements in place of if-elseif-blocks forcing handling of all spool-state cases
  - why do stabilises, acros and drift not handle spool states of spool-up and spool-down for the purposes of resetting integrators and the like? e.g. your motors really should leave their lower limit in spool-up and definitely be off the limit in unlimited!
- quadplane's assertion that motors are shut down seems to be incorrect here: https://github.com/ardupilot/ardupilot/blob/master/ArduPlane/quadplane.cpp#L1705 as they may be in spool-up

### CODE REVIEW

File: libraries/AP_Motors/AP_Motors_Class.h [hunk 1]
```diff
@@ -221,8 +221,8 @@ class AP_Motors {
     float               _lateral_in;                // last lateral input from set_lateral caller
     float               _throttle_avg_max;          // last throttle input from set_throttle_avg_max
     LowPassFilterFloat  _throttle_filter;           // throttle input filter
-    spool_up_down_desired _spool_desired;           // desired spool state
-    spool_up_down_mode  _spool_mode;                // current spool mode
+    DesiredSpoolState _spool_desired;           // desired spool state
+    SpoolState  _spool_state;                // current spool mode
```
Comment (rmackay9) [inline libraries/AP_Motors/AP_Motors_Class.h]: small indentation issue?  maybe it's just the web viewer..

File: ArduCopter/mode.cpp [hunk 2]
```diff
@@ -410,18 +410,20 @@ void Copter::Mode::zero_throttle_and_hold_attitude()
 
 void Copter::Mode::make_safe_spool_down()
 {
-    // command aircraft to initiate the shutdown process
-    motors->set_desired_spool_state(AP_Motors::DESIRED_GROUND_IDLE);
-    switch (motors->get_spool_mode()) {
+    // command aircraft to initiate the shutdowon process
```
Comment (rmackay9) [inline ArduCopter/mode.cpp]: added a typo on "shutdown"

File: ArduCopter/mode.cpp [hunk 3]
```diff
@@ -627,15 +629,15 @@ Copter::Mode::AltHoldModeState Copter::Mode::get_alt_hold_state(float target_cli
...
         // transition through states as aircraft spools down
-        switch (motors->get_spool_mode()) {
+        switch (motors->get_spool_state()) {
 
-        case AP_Motors::SHUT_DOWN:
+        case AP_Motors::SpoolState::SHUT_DOWN:
             return AltHold_MotorStopped;
 
-        case AP_Motors::DESIRED_GROUND_IDLE:
+        case AP_Motors::SpoolState::GROUND_IDLE:
```
Comment (rmackay9) [inline ArduCopter/mode.cpp]: So this was using the wrong enum but luckily "DESIRED_GROUND_IDLE" and "GROUND_IDLE" were both enumerated as '1".

File: ArduCopter/mode_acro.cpp [hunk 4]
```diff
@@ -16,27 +16,35 @@ void Copter::ModeAcro::run()
...
+        break;
+    case AP_Motors::SpoolState::THROTTLE_UNLIMITED:
         // clear landing flag above zero throttle
         if (!motors->limit.throttle_lower) {
             set_land_complete(false);
         }
+        break;
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing and hope for the best!
```
Comment (rmackay9) [inline ArduCopter/mode_acro.cpp]: maybe we can shorten this comment to just "// do nothing".  I think we don't need to do anything during spool up or spool down.

File: ArduCopter/mode_acro_heli.cpp [hunk 5]
```diff
@@ -36,27 +36,35 @@ void Copter::ModeAcro_Heli::run()
...
+        break;
+    case AP_Motors::SpoolState::THROTTLE_UNLIMITED:
         // clear landing flag above zero throttle
         if (!motors->limit.throttle_lower) {
             set_land_complete(false);
         }
+        break;
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing and hope for the best!
```
Comment (rmackay9) [inline ArduCopter/mode_acro_heli.cpp]: here too, let's shorten this comment to "// do nothing"

File: ArduCopter/mode_auto.cpp [hunk 6]
```diff
@@ -1729,7 +1729,7 @@ bool Copter::ModeAuto::verify_RTL()
 {
     return (copter.mode_rtl.state_complete() && 
     (copter.mode_rtl.state() == RTL_FinalDescent || copter.mode_rtl.state() == RTL_Land) &&
-    (motors->get_spool_mode() == AP_Motors::GROUND_IDLE));
+            (motors->get_spool_state() == AP_Motors::SpoolState::GROUND_IDLE));
```
Comment (rmackay9) [inline ArduCopter/mode_auto.cpp]: missing indentation on the line above?

File: ArduCopter/mode_drift.cpp [hunk 7]
```diff
@@ -80,27 +80,35 @@ void Copter::ModeDrift::run()
...
+        break;
+    case AP_Motors::SpoolState::THROTTLE_UNLIMITED:
         // clear landing flag above zero throttle
         if (!motors->limit.throttle_lower) {
             set_land_complete(false);
         }
+        break;
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing and hope for the best!
```
Comment (rmackay9) [inline ArduCopter/mode_drift.cpp]: let's shorten this comment as well to just, "// do nothing"

File: ArduCopter/mode_stabilize.cpp [hunk 8]
```diff
@@ -20,27 +20,35 @@ void Copter::ModeStabilize::run()
...
+        break;
+    case AP_Motors::SpoolState::THROTTLE_UNLIMITED:
         // clear landing flag above zero throttle
         if (!motors->limit.throttle_lower) {
             set_land_complete(false);
         }
+        break;
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing and hope for the best!
```
Comment (rmackay9) [inline ArduCopter/mode_stabilize.cpp]: ditto here..

File: ArduCopter/mode_stabilize_heli.cpp [hunk 9]
```diff
@@ -41,27 +41,34 @@ void Copter::ModeStabilize_Heli::run()
...
-    } else if (motors->get_spool_mode() == AP_Motors::THROTTLE_UNLIMITED) {
+        break;
+    case AP_Motors::SpoolState::THROTTLE_UNLIMITED:
         // clear landing flag above zero throttle
         if (!motors->limit.throttle_lower) {
             set_land_complete(false);
         }
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing and hope for the best!
```
Comment (rmackay9) [inline ArduCopter/mode_stabilize_heli.cpp]: ditto here ..

**Comment by rmackay9** [comment]
I like this change because it fixes the formatting of the num name and ensures that we don't mixup desired-spool-state with actual-spool-state.

@bnsgeyer, are you happy with the new naming?  It's a little bit longer but it was already pretty long.

File: ArduCopter/mode_acro.cpp (hunk 4 above)
Comment (peterbarker) [reply ArduCopter/mode_acro.cpp]: That makes me very curious, actually, and the reason I changed things to be switch statements.  In these states you should be on the ground - but we're not resetting targets and the like.

Needs attention from a controls person - if only to add a comment.

In the meantime I'll shorten it to "do nothing"

File: libraries/AP_Motors/AP_Motors_Class.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Motors/AP_Motors_Class.h]: Fixed.

File: ArduCopter/mode.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/mode.cpp]: Fixed.

File: ArduCopter/mode.cpp (hunk 3 above)
Comment (peterbarker) [reply ArduCopter/mode.cpp]: Yes.

File: ArduCopter/mode_acro.cpp (hunk 4 above)
Comment (peterbarker) [reply ArduCopter/mode_acro.cpp]: Shortened all of these to just "do nothing"

File: ArduCopter/mode_auto.cpp (hunk 6 above)
Comment (peterbarker) [reply ArduCopter/mode_auto.cpp]: Removed whitespace change.

**Comment by bnsgeyer** [comment]
@peterbarker and @rmackay9   Give me the weekend to get to this.  I don't have time to look at this right now.  Thanks,
Bill

**Comment by bnsgeyer** [comment]
@rmackay9 I am fine with these changes.

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request merged**